[24기 이송현] Add: 메인 페이지 컴포넌트 리뷰 요청

vi2920va opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 메인 페이지 레이아웃 완성
-  ul에 list MockData로 데이터 완성 후 메인해서 필터링해주는 로직 구현
 - nextBtn 버튼 클릭시 다음 이미지 슬라이드 로직 구현
 - prevBtn 버튼 클릭시 이전 이미지 슬라이드 로직 구현
- 상단/하단 캐러셀에서 이미지 이동 할 때 보여주는 이미지 번호 바꾸는 로직 구현
- 캐러셀 이미지에서 첫 번째 이미지 또는 마지막 이미지 이면 버튼 비활성화 기능
- main.scss에서 구글 웹 폰트 적용
![메인페이지 이미지](https://i.postimg.cc/sfKMWbQ2/screencapture-localhost-3002-2021-09-06-17-46-22.png)
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 메인 페이지 리스트 아이템 컴포넌트로 분리
- 메인 페이지 슬라이드쇼 marigin state가 아닌 변수로 값을 변경
- 404 경로 수정 및 마크업 수정
- mainList아이템 불필요한 font 스타일 제거
- mockData 아이템 id number로 변경
- componentDidMount 불러오는 라이프 사이클 메서드에서 변수명 변경 
- componentDidMount forEach 객체 분리



## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)



## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

vi2920va committed:
Add: 메인 페이지 컴포넌트 생성

vi2920va committed:
Modify: 메인 페이지 컴포넌트 레이아웃 완료

vi2920va committed:
Modify: Router 컴포넌트 MainPage  연결

vi2920va committed:
Modify: resetStyle 수정
- a color 초기화

vi2920va committed:
Modify: 메인 페이지 캐러셀 기능 구현
- saleList MockData 연결 로직 구현
- nextBtn 버튼 클릭시 다음 이미지 슬라이드 로직 구현
- prevBtn 버튼 클릭시 이전 이미지 슬라이드 로직 구현

vi2920va committed:
Add: 메인 페이지 연결 mockData
- 메인 페이지 컴포넌트 리스트에 연결하기 위한 데이터

vi2920va committed:
Modify: Main Mock Data 추가
- 카테고리 용도별로 슬라이드쇼 하기 위해서 데이터 추가

vi2920va committed:
Refactor: 슬라이드쇼 셀프 리팩토링
- saleList 슬라이드쇼 셀프 리팩토링
- insideList 슬라이드쇼 로직 구현중
- error: insideList는saleList 처럼 슬라이드쇼를 구현할 경우
 	 content 내용을 짤림

vi2920va committed:
Refactor: 슬라이드쇼 셀프 리팩토링
- saleList, insideList 각각의 슬라이드쇼 로직 구현

vi2920va committed:
Modify: 메인 페이지 수정

vi2920va committed:
Modify: 데이터 수정

vi2920va committed:
Merge branch 'main' into feature/main

vi2920va committed:
Refactor: 레이아웃 셀프 리팩토링
- newProductWrapper background-img로 스타일 및 마크업 변경
- neWCollectionWrapper background-img로 스타일 및 마크업 변경
- lafestWorldWrapper 마크업 변경 및 레이아웃 스타일 변경
- lafestWorldWrapper MockData 적용예정
- lafestInfestInsideWrapper 레이아웃 변경 예정

vi2920va committed:
Modify: Data 변경
- 사용하지 않는 의미 없는 id 삭제

vi2920va committed:
Modify: MockData id 부여
- 각각의 아이템별로 고유한 id값 부여

vi2920va committed:
Modify: 메인 페이지 상세 레이아웃
- 상단과 하단의 슬라이드쇼 로직 구현
- 슬라이드쇼 이미지 갯수 로직 구현
- ul 태그 마크업 mockData로 만들어서 카테고리에 맞는 배열 필터링하기 로직 구현
- 구글 웹 폰트 main.scss에서import 파일 적용

vi2920va committed:
Modify: 메인페이지 불필요한 주석 제거

saengmotmi commented on public/data/mainData-songhyun.json:
```diff
@@ -0,0 +1,91 @@
+[{
+    "id": "mainEvent1",
```
id는 보통 db에서 데이터의 pk 값으로 잡기 때문에 number 형태로 넘어올거 같습니다. 백엔드와 같이 데이터 구조 확인부탁드립니다

vi2920va replied:
종택님 저희가 메인페이지에서 보여주는 이미지 파일들을 백엔드와 연결을 구현하지 않아서 이 부분은 백엔드와 데이터 구조 확인이 어렵습니다 😅

yeonuk-hwang replied:
나중에도 백엔드와 연결은 하지 않고 프론트 내부에서 목데이터 형태로 구현한다는 말씀이신가요? 그래도 id라는 값은 일반적으로 number 형태로 저장됩니다. 그래서 프론트 내부적으로 처리하더라도 id값을 number 형식으로 맞춰주시는게 좋아보이네요!

saengmotmi commented on src/Routes.js:
```diff
@@ -1,32 +1,32 @@
           <Route exact path="/productList" component={ProductList} />
           <Route exact path="/signUp" component={SignUp} />
           <Route exact path="/logIn" component={Login} />
+          <Route export path="/error" component={NotFound} />
```
보통 not found 페이지 같은 경우 exact path에 해당되지 않은 나머지 전체 경로에 대해 /error 페이지로 리다이렉트 시켜주는 로직으로 구성합니다. react-router 공식 문서에서 redirect 컴포넌트 검색해보세요!

vi2920va replied:
넵 알겠습니다!

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,240 @@
+    const { saleCount, saleMargin, saleList } = this.state;
+
+    if (!saleCount < saleList.length && saleCount > 1) {
+      this.setState({ saleCount: saleCount - 1 });
+      this.setState({ saleMargin: saleMargin + 421 });
```
```suggestion
      this.setState({
        saleCount: saleCount - 1,
        saleMargin: saleMargin + 421
      });
```
객체 형태이므로 한번에 업데이트 해주시면 됩니다

vi2920va replied:
넵 한 번에 업데이트 형태로 바꿨습니다. 그리고marigin은 state로 관리하지 않고 따로 변수에서 값을 관리하는 걸로 바꿧습니다.

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,240 @@
+    fetch('http://localhost:3002/data/mainData-songhyun.json')
+      .then(res => res.json())
+      .then(items => {
+        this.setState({
+          saleList: items.filter(item => item.category.includes('event')),
+          insideList: items.filter(item => item.category.includes('inside')),
+          worldList: items.filter(item => item.category.includes('collection')),
+        });
```
이 부분은 filter를 각각 세번 돌리는 게 아니라 가져온 데이터에 대해 루프를 한번 돌면서 최종적으로 필요한 객체 형태를 만들어주는게 좋지 않을까요?

vi2920va replied:
이 부분 수정해서 하나의 배열로 map을 사용해서 변경했습니다

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,240 @@
+          </div>
+          <ul className="insides" style={insideStyle}>
+            {insideList.map(insideItem => (
+              <li key={insideItem.id}>
+                <Link to="#">
+                  <figure>
+                    <img src={insideItem.url} alt={insideItem.title} />
+                    <figcaption className="insideContent">
+                      <span>{insideItem.subtitle}</span>
+                      <h3>{insideItem.title}</h3>
+                      <strong>{insideItem.content}</strong>
+                      <button type="button" className="summaryBtn">
+                        자세히 보기
+                      </button>
+                    </figcaption>
+                  </figure>
+                </Link>
+              </li>
```
이런식으로 반복되는 부분은 충분히 컴포넌트로 나눠볼 수 있을거 같습니다

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,240 @@
+                  type="button"
+                  className="nextBtn"
+                  onClick={this.handleInSideNextClick}
+                  disabled={insideCount === insideList.length && true}
```
```suggestion
                  disabled={insideCount === insideList.length}
```
연산 결과로 떨어진 boolean 값을 그대로 이용해도 되겠네요

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,240 @@
+        <section className="lafestWorldWrapper">
+          <h2>The Lacoste World</h2>
+          <ul className="lafeWorlds">
+            {worldList.map(item => (
+              <li key={item.id}>
+                <Link to="#">
+                  <figure>
+                    <img src={item.url} alt={item.title} />
+                    <figcaption className="lafestWorlContent">
+                      <h3>{item.title}</h3>
+                      <strong>신상품 7% 혜택</strong>
+                      <button type="button">구매하기</button>
+                    </figcaption>
+                  </figure>
+                </Link>
+              </li>
+            ))}
```
마찬가지입니다

saengmotmi commented on src/pages/MainPage/MainPage.scss:
```diff
@@ -0,0 +1,356 @@
+      margin-top: 4.5%;
+      margin-bottom: 1.5%;
+
+      h3 {
+        font-family: 'Noto Sans KR', sans-serif;
+        font-size: 19px;
+        font-weight: 600;
+      }
```
이 부분만 noto sans 적용되는 건가요?

vi2920va replied:
메인에서 따로 폰트 적용되는 부분입니다

saengmotmi commented on src/pages/MainPage/MainPage.scss:
```diff
@@ -0,0 +1,356 @@
+
+  .newProductSaleWrapper {
+    position: relative;
+    display: flex;
+    flex-direction: column;
```
자식요소들이 전부 block요소라면 굳이 flex로 세로 배열 할 필요 없을거 같습니다

vi2920va replied:
넵 이 부분 수정했습니다

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,240 @@
+    this.state = {
+      saleCount: 1,
+      insideCount: 1,
+      saleMargin: 0,
+      insideMargin: 0,
```
혹시 margin 값이 count에 따라 연동되는 거라면 단위(421)을 상수 변수로 분리하고, 해당 변수와 count를 연산하는 식으로 활용할 수 있을거 같아요. margin을 state에 저장하지 않아도 된다는 뜻입니다

vi2920va replied:
> @vi2920va 충돌 해결하신거 아직 push가 안된거 같아요! 확인해주세요

넵충돌 해결 했습니다.

saengmotmi reviewed (commented):
코멘트 확인해주세요!

vi2920va committed:
Refactor: 메인 페이지 리팩토링
- 메인 페이지 리스트 아이템 컴포넌트로 분리
- 슬라이드쇼 이동 marigin state로 관리하지 않고 따로 변수에 값을 관리하는 로직 분리
- 메인 페이지 리스트 하나의 배열 리스트로 관리

vi2920va committed:
Modify: Routes 404페이지 경로 수정

vi2920va committed:
Modify: 메인페이지 MockData(content) 내용 추가

vi2920va committed:
Modify: Routes 404 페이지 경로 변경

vi2920va committed:
Merge branch 'main' into feature/main

saengmotmi commented:
@vi2920va 충돌 해결하신거 아직 push가 안된거 같아요! 확인해주세요

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,227 @@
+            </ol>
+          </div>
+          <ul className="insides" style={insideStyle}>
+            {list.map(item => {
+              if (item.category === 'inside') {
+                return (
+                  <>
+                    <MainItem item={item} />
+                  </>
+                );
+              }
+            })}
```
map 말고 filter로 처리하면 더 좋겠군요

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,227 @@
+        <section className="lafestWorldWrapper">
+          <h2>The Lacoste World</h2>
+          <ul className="lafeWorlds">
+            {list.map(item => {
+              if (item.category === 'collection') {
+                return (
+                  <>
+                    <MainItem item={item} />
+                  </>
+                );
+              }
+            })}
+          </ul>
```
마찬가지이고 fragment 불필요해보입니다
key props도 넣어주세요

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,227 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      evnentCount: 1,
```
```suggestion
      eventCount: 1,
```
오타 수정해주세요

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,227 @@
+import MainItem from './MainItem/MainItem';
 import './MainPage.scss';
 
+let saleMargin = 0;
+let insideMargin = 0;
```
이런식으로 사용하는 건 가급적 없어야겠습니다. 현재 구조상 그럴 일이 없긴 하지만 컴포넌트가 여러번 불릴 경우 오작동 하게 될 코드이고, 차라리 컴포넌트의 멤버 변수를 선언하시는게 좋겠습니다.

saengmotmi commented on src/pages/MainPage/MainItem/MainItem.js:
```diff
@@ -0,0 +1,27 @@
+    const { item } = this.props;
+
+    return (
+      <li key={item.id}>
```
key는 반복문을 사용하는 측에서만 있으면 됩니다 여기는 있어도 의미가 없어요

saengmotmi commented on src/pages/MainPage/MainItem/MainItem.js:
```diff
@@ -0,0 +1,27 @@
+import React, { Component } from 'react';
+import { Link } from 'react-router-dom';
+import './MainItem.scss';
+class MainItem extends Component {
```
```suggestion
import './MainItem.scss';

class MainItem extends Component {
```
구분되는 요소는 한줄 띄워주세요

vi2920va committed:
Fix: confilct 충돌 해결

saengmotmi commented on src/Routes.js:
```diff
@@ -1,28 +1,29 @@
           <Route exact path="/productList" component={ProductList} />
           <Route exact path="/signUp" component={SignUp} />
           <Route exact path="/logIn" component={Login} />
+          <Route path="*" component={NotFound} />
```
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 호다닥 적용하셨네요

vi2920va replied:
종택님 저 리뷰 반영했습니다!

saengmotmi reviewed (commented):
리뷰 확인해주세요

vi2920va committed:
Fix: 충돌 해결 및 리뷰 반영

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,230 @@
+          };
+          init[obj[item.category]].push(item);
+        });
+        this.setState({ list: [init] });
```
```suggestion
        this.setState({
          insideList: [...],
          eventList: [...],
          collectionList: [...],
        });
```
이런 형태로 될 수 있도록 다 만들어두셨잖아요...ㅠㅠ state에 있는 list를 3종류의 state로 분리하라는 뜻이었습니다

vi2920va replied:
뜨끔....종택님 새벽에 안 주무셨어요? 리뷰반영하여 수정했습니다!

yeonuk-hwang replied:
반영하신 것 push 안된 것 같네요

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,230 @@
+            </ol>
+          </div>
+          <ul className="newProducts" style={saleStyle}>
+            {list.map(item =>
+              item.eventList.map(item => <MainItem key={item.id} item={item} />)
```
그렇게 되면 이중 맵 돌릴 필요가 없습니다
```suggestion
            {eventList.map(item => <MainItem key={item.id} item={item} />)
```

saengmotmi commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,230 @@
+              <li className="endCount">
+                <span>{Number(insideLength)}</span>
+              </li>
+              <li>
+                {
+                  <button
+                    type="button"
```
```suggestion
              <li>
                  <button
                    type="button"
```
중괄호가 불필요해보입니다

yeonuk-hwang replied:
반영해주세요

saengmotmi commented on src/pages/MainPage/MainPage.scss:
```diff
@@ -0,0 +1,215 @@
+    border-bottom: 1px solid #e5e5e5;
+    background: #f4f4f4;
+
+    div {
```
div 태그 광범위하게 적용될 수 있어서 주의해서 사용하셔야 합니다. 클래스 셀렉터 사용해서 특정해주세요.

yeonuk-hwang replied:
반영안된부분 다 반영해주세요

saengmotmi commented on src/pages/MainPage/MainPage.scss:
```diff
@@ -0,0 +1,215 @@
+  }
+
+
+  .lafestInsideWrapper {
+    display: flex;
+    flex-direction: column;
```
세로 정렬하는거 뿐이라면 display block 속성 사용이 더 적합해보입니다

vi2920va committed:
Modify: 메인 페이지 리뷰 반영 수정

yeonuk-hwang commented on src/pages/Error/NotFound.js:
```diff
@@ -0,0 +1,20 @@
+class NotFound extends Component {
+  render() {
+    return (
+      <div className="notFoundContainer">
```
h1~h6등 해당 구획의 제목요소들을 감싸는 태그로는 div보다는 header tag 가 적합해보입니다!

yeonuk-hwang commented on src/pages/Error/NotFound.scss:
```diff
@@ -0,0 +1,30 @@
+.notFoundContainer {
+  display: flex;
+  flex-direction: column;
+  text-align: center;
+  justify-content: center;
```
자주사용되는 속성들 mixin으로 활용해보시는것도 좋아보이네요

yeonuk-hwang commented on src/pages/Error/NotFound.scss:
```diff
@@ -0,0 +1,30 @@
+    margin-bottom: 20px;
+  }
+
+  p {
+    &:last-of-type {
+      margin-bottom: 20px;
+    }
+  }
```
```suggestion
  p:last-of-type {
      margin-bottom: 20px;
  }
```
이렇게 적용하는것과 차이가 없어보입니다!

yeonuk-hwang commented on src/pages/MainPage/MainItem/MainItem.js:
```diff
@@ -0,0 +1,27 @@
+
+class MainItem extends Component {
+  render() {
+    const { item } = this.props;
```
`const {url, title, category} = itme`
한번 더 구조분해할당해서 써도 되 보입니다.

vi2920va replied:
한 번더 구조분해 할게요!

yeonuk-hwang commented on src/pages/MainPage/MainItem/MainItem.scss:
```diff
@@ -0,0 +1,143 @@
+@import url('https://fonts.googleapis.com/css2?family=Noto+Sans+KR:wght@700&display=swap');
+@import url('https://fonts.googleapis.com/css2?family=Rubik:wght@500&display=swap');
```
mainItem에는 className이 부여된 곳이 figcaption에 동적으로 들어가는 부분밖에 없어보이는데 그 외의 스타일링도 이 파일에 포함되어 있어보이네요 이런식으로 css를 작성하면 컴포넌트별로 독립적으로 분리해서 사용하기 힘들어집니다 컴포넌트의 scss파일에서는 해당 컴포넌트의 스타일링만 책임져주세요

yeonuk-hwang commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,229 @@
+
+  handleNextClick = () => {
+    const { eventCount, list } = this.state;
+    const length = Number(list.map(item => item.eventList.length));
```
map의 리턴값은 배열인데 그걸 Number로 형변환 하면 NaN이 출력될 것 같습니다. 제대로 동작되는지 확인해보세요!

yeonuk-hwang commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,229 @@
+  handleNextClick = () => {
+    const { eventCount, list } = this.state;
+    const length = Number(list.map(item => item.eventList.length));
+    console.log(length);
```
테스트 마친 콘솔은 제거해주세요

yeonuk-hwang commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,229 @@
+  };
+
+  componentDidMount() {
+    fetch('http://localhost:3007/data/mainData-songhyun.json')
```
```suggestion
    fetch('/data/mainData-songhyun.json')
```
localhost:3007 등의 현재 구동되고 있는 서버의 주소는 안적어주셔도 됩니다. 절대경로로 적으면 CRA 내부적으로 자동으로 현재 구동되고 있는 서버 주소를 넣어줍니다, 오히려 이렇게 임의적으로 기입하게 되면 3007포트가 아닌 3000포트 등에서 돌아갈 때는 에러가 발생하겠네요

yeonuk-hwang commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,229 @@
+        };
+
+        lists.forEach(item => {
+          const obj = {
+            inside: 'insideList',
+            collection: 'collectionList',
+            event: 'eventList',
+          };
```
forEach loop 안에서 obj를 선언하는 연산을 하게 되면 매 루프마다 obj를 선언하고 값을 할당하게 됩니다. 한번만 선언되고 할당되어도 되는 객체로 보이니까 forEach loop 바깥으로 분리하면 불필요한 연산을 제거할 수 있겠네요

yeonuk-hwang replied:
변수명도 obj보다는 명확히 의미가 드러나는 것으로 변경해주세요

vi2920va replied:
넵 이 부분 적용하겠습니다

yeonuk-hwang commented on src/pages/MainPage/MainPage.js:
```diff
@@ -1,9 +1,229 @@
+                <span>/</span>
+              </li>
+              <li className="endCount">
+                <span>{Number(saleLength)}</span>
```
saleLength는 배열인 것 같네요 위에서 드린 코멘트와 동일합니다.

yeonuk-hwang reviewed (commented):
리뷰 확인 및 반영해주세요 송현님

vi2920va committed:
Modify: 메인 컴포넌트 리뷰 반영

vi2920va committed:
Merge branch 'main' into feature/main

vi2920va committed:
Modify: 메인 컴포넌트 주석 및 레이아웃 스타일 수정

vi2920va committed:
Modify: 가독성을 위한 코드 분리

saengmotmi commented:
@vi2920va 고생하셨습니다 🔥
